src/Microsoft.AspNetCore.Authorization.Middleware/AuthorizationMiddleware.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authorization.Policy;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Authorization
+{
+    public class AuthorizationMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly IAuthorizationPolicyProvider _policyProvider;
+
+        public AuthorizationMiddleware(RequestDelegate next, IAuthorizationPolicyProvider policyProvider)
+        {
+            if (next == null)
+            {
+                throw new ArgumentNullException(nameof(next));
+            }
+
+            if (policyProvider == null)
+            {
+                throw new ArgumentNullException(nameof(policyProvider));
+            }
+
+            _next = next;
+            _policyProvider = policyProvider;
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var endpoint = context.Features.Get<IEndpointFeature>()?.Endpoint;
+            if (endpoint == null)
+            {
+                await _next(context);
+                return;
+            }
+
+            var policyEvaluator = context.RequestServices.GetRequiredService<IPolicyEvaluator>();
+
+            // REVIEW: Does order matter? Is ordered metadata in the correct order?
+            var authorizeData = endpoint.Metadata.GetOrderedMetadata<IAuthorizeData>();
```

Comment:
I'll let @HaoK respond as to whether order matters when combining policies. `GetOrderedMetadata` returns items from most specific to least specific.

Reply:
Order generally doesn't matter more combining policies with the default implementations since all requirements in all policies must be satisfied.  It might affect the order in which authentication builds the claims principal as well, but that shouldn't be order sensitive for the resulting claim set, just you would see slight differences in the order of execution, so I think its fine.